### ISSUE TITLE
MarkLogic DataHub and Unit Test

### ISSUE THREAD

**Issue by chenfiona** [marklogic-community/marklogic-unit-test#124]
Hi: I need some insight as how to incorporate Unit Test within Data Hub. Now, I have finished the DH flows, harmonisation, mastering. I need some clarity:
1) Is the Unit Test applicable in a DH ?  If yes, what does the Unit Test test exactly? Like data quality, module function or else ? 
2) It seems to me that the example data is just couple documents. I have thousands of documents in the DH. Should I configure the Unit Test pointing to the Staging and Final database? Is the Unit Test build.gradle/gradle.properties completely separately from the DH build.gradle/gradle.properties?

**Comment by dmcassel** [comment]
1. Is the Unit Test applicable in a DH ?

Yes. Note that the standard deployment trick of just adding `mlTestRestPort` won't do it, but see [this Gist](https://gist.github.com/rjrudin/0dd97eaee16b1f8dc75a35c7acf7b4e1) for tasks you can add to your `build.gradle`. 

2. If yes, what does the Unit Test test exactly? Like data quality, module function or else ?

marklogic-unit-test is an effective way to test individual functions or HTTP endpoints. If your DHF flows rely on custom modules, those custom modules are good candidates. If you have REST extensions or other custom code for serving data, those are good candidates, too. 

3. seems to me that the example data is just couple documents. I have thousands of documents in the DH. Should I configure the Unit Test pointing to the Staging and Final database? Is the Unit Test build.gradle/gradle.properties completely separately from the DH build.gradle/gradle.properties?

See the gist I referred to above. You'll want a separate app server and content database for your tests. The reason for this is that, while testing, you want to have control over what content the tests are seeing. The tests will do setup (which often add some test data), run tests, and teardown (which deletes data). The build.gradle file won't be separate, but it will create separate resources for testing.

**Comment by chenfiona** [comment]
Thanks! Is there any relationship among below 4 ports?

( These 2 ports are in build.gradle. I set them as 8034 and 8035)
-> mlStagingAppserverName + "-TEST", 8034
-> mlFinalAppserverName + "-TEST", 8035

(These 2 ports are in gradle.properties.)
=> mlRestPort=8003
=> mlTestRestPort=8004

**Comment by dmcassel** [comment]
I believe the latter two ports are not used with DHF projects

**Comment by chenfiona** [comment]
build.gradle and mlBundle structure snippet:
[Unit Test Screenshot.zip](https://github.com/marklogic-community/marklogic-unit-test/files/4366523/Unit.Test.Screenshot.zip)
gradle.properties: ( just add one line at the end) 
mlModulePaths=src/main/ml-modules,src/test/ml-modules

After mlDeploy/mlRedeploy ran successfully, I don't see the mlStagingAppserverName + "-TEST", 8034 or mlFinalAppserverName + "-TEST", 8035 through Admin console.  
I attached the build.gradle.  Can you identify what is missing?

**Comment by chenfiona** [comment]
@David: I managed to create the TEST DBs ( one for staging, another for final ) and corresponding App servers ( one for staging, another for final ). 
But Paxton's version has conflict with rjrudin's: 
Paxton's **hubDeployTestServers** contains 5 parameters but **DeployHubTestServerCommand** only accepts 4.
It throws error:

*** What went wrong:
Execution failed for task ':hubDeployTestServers'.
> Could not find matching constructor for: DeployHubTestServerCommand(com.marklogic.hub.impl.DataHubImpl, String, String, Integer, String)**

So I use  rjrudin's.  
Below is the Bundle snippet:
![Unit-Test_DBs](https://user-images.githubusercontent.com/53400797/77270926-a8661780-6c83-11ea-87bd-3f092f46021b.jpg)

Will continue tomorrow !

**Comment by chenfiona** [comment]
So I attempt to deploy the TEST resources in another environment ( this is DH 5.1.0, the above is DH 5.2.0) , but got below error ( the TEST DBs are deployed but the App servers deployment failed)

*** Where:
Build file 'C:\mldhf\build.gradle' line: 120
* What went wrong:
Execution failed for task ':hubDeployTestServers'.
> Index 0 out of bounds for length 0**

The failed process is below:

  DeployHubTestServerCommand(String serverFilenamePattern, String serverName, int port, String contentDatabaseName) {
(line 120)        **super()**
        setResourceFilenamesIncludePattern(Pattern.compile(serverFilenamePattern))
        this.serverName = serverName
        this.port = port
        this.contentDatabaseName = contentDatabaseName
    }

**Comment by chenfiona** [comment]
> Task :hubDeployTestFinalDatabase
Caching disabled for task ':hubDeployTestFinalDatabase' because:
  Build cache is disabled
Task ':hubDeployTestFinalDatabase' is not up-to-date because:
  Task has not declared any outputs despite executing actions.

**Comment by chenfiona** [comment]
Hi, David:
            I have no option but upgrading our DH to 5.2.0.

**Comment by chenfiona** [comment]
After the Test Resources & App Servers are deployed in DH, should I run below or something else?
./gradlew mlGenerateUnitTestSuite

**Comment by chenfiona** [comment]
I generated the test suite and Run Tests in the Final Test REST API.  It shows: 
***-hub-FINAL-TEST Unit Tests: all passed

According to the Watch, test module timestamps indicate few scan through the custom modules and the Rest extensions sjs :
\\src\\test\\ml-modules\\root\\test\\suites\\sampletestsuite\\****

What does the "all passed" mean exactly?

**Comment by dmcassel** [comment]
I believe "all passed" means that it found one or more assertions and that those assertions we successful. I don't think you'd get that if it didn't find any assertions.

**Comment by chenfiona** [comment]
Thanks!  Are below options doable in DH Unit Test?
1) Fine-grained test suites produced in DH. Out of DH custom modules and REST extensions, I want to overwrite SampleTestSuite. Assuming the branch suite path is /test/suites/, name is productTest. When I run UnitTest , it should test productTest which tests the REST extension called prodSearch.sjs. ( I will pass the q to the REST and returns the query result. )
If yes, is there example of such use case?

2) Data quality in DH:  As document is ingested into DH ( example DMSDK), I want to assertEqual the expected vs actual property value.
If yes, is there example of such use case?

**Comment by dmcassel** [comment]
@chenfiona it sounds like you've worked through some changes needed for DHF 5.2. Would you mind posting what you have? 

1. I like to move most of the work from my REST extensions into a library. That way I can import the library into a test case and focus on the logic. The REST extension's job is just to convert the HTTP parameters to function parameters. So in my approach, prodSearch.sjs would have just a thin layer and defer most of the work to searchLib.sjs (or whatever you'd like to call it). I'd set up /test/suites/searchLib/ and build test cases to make sure my functions worked as expected. If there was a lot in searchLib.sjs, it might even be worth splitting into multiple test suites. 

If the conversion from HTTP parameters to function parameters is non-trivial, then you can use the HTTP test functions and perhaps a secure credential to test that level. 

2. For testing, your suiteSetup.sjs or setup.sjs is responsible for loading your test data. Are you talking about doing an assertion during regular run time (ie, not during testing)?

**Comment by chenfiona** [comment]
[build.zip](https://github.com/marklogic-community/marklogic-unit-test/files/4383677/build.zip)
1. 
I am most glad to share the build.gradle as above attachment. It works for DH 5.2.0 ONLY.
The gradle.properties is updated with below line:
mlModulePaths=src/main/ml-modules,src/test/ml-modules

I leave that to you as how to use the Unit Test to test DH specific modules(main.sjs)/REST endpoint. This Unit Test should use existing DH data. 

2. Please review below snippet: 
![test-ml-modules](https://user-images.githubusercontent.com/53400797/77588874-b05acd00-6eb8-11ea-819c-72a6a392e9b3.JPG)
  In the MarkLogic Unit Test introduction video, there is setup.sjs ( load some test data), teardown.sjs, {test-module}.sjs.  If user loads some data into DH Staging, I would expect to Assert the data ( example: the unitPrice expected value is 500.9998 but the actual loaded value is 500.99. Then assertEqual should fail. 
In DH, the data load is a regular real-time operation. If the Unit test passes for few testing data, you would expect that data load logic works for other data?
I may need clarification about this Unit Test: One of the expectations comes from the Roxy.  My understanding is that Roxy Unit Test is replaced by MarkLogic Unit Test.  Isn't the Unit Test testing of csv convert in your Roxy: Unit testing video?

**Comment by dmcassel** [comment]
Thanks for uploading! 

Yes, what used to be Roxy Unit Test is now MarkLogic Unit Test. It's a separate project so that it can be easily integrated with ml-gradle projects, without any dependency on Roxy. 

I think you're saying that you want to run marklogic-unit-test against real data in your real database. You could do that, but know that it's not the intended use of the framework. (If I've misread your goals, my apologies and please correct me.) To map what I'm reading of your case to the intended way to use the framework -- how is the data loaded into DH Staging? What is it you're really trying to test? Is there a transform or a library with some validation steps? I would build a test that has some valid test data and test that my transform or validation successfully recognizes it as good. I'd also set up test data that had a variety of invalid values and test that those were recognized as bad. 

In other words, I'd use MarkLogic Unit Test to test the logic that I'm using to validate data, rather than to directly validate my actual data.

**Comment by chenfiona** [comment]
Yes, you are right. We have an internal discussion regarding MarkLogic Unit Test. 

Background of Data Hub:  DMSDK is already in-house to load data into ML. We would like to continue this ingestion approach into DH Staging. We would like to utilize ML DH to map, customize/harmonise, match/merge and serve that greatest data.

[Unit Test.zip](https://github.com/marklogic-community/marklogic-unit-test/files/4388584/Unit.Test.zip)

Desired Unit Test Usage:

1)	Please review uploaded [**mapping.test.js**] 
-	Line 1 imports the testing module: we are testing DH mapping logic 
-	Line 2 should be the MarkLogic Unit Test util module
-	Line 4 indicates a Data Hub mock unit test
-	Line 11- 23 is the origin instance; Line 31-49 is the mapped entity 

If **LocalValue: "lv"** is returned something else like "lvl", then toEqual/toBe/assertEqual/deepEqual should fail (the mapping logic).

We would like to test other server-side JS logic in DH using the MarkLogic Unit Test: Example a custom main.sjs for the Account flow

2)	Please review uploaded [account custom.doc] as another test data structure
-	Line 1-6 is the document towards customization in DH
-	Line 7-13 is the harmonized document using a custom module. There is complex business logic during data transformation.

If **"accountNumber": "A10"** is returned something else like "G6", then toEqual/toBe/assertEqual/deepEqual should fail (custom module logic).

Same token, if "accountTypeDescription": "Mary Jane" is returned something else (note the value is obtained through other lookup table), then toEqual/toBe/assertEqual/deepEqual should fail (custom module logic).

The only difference between 1) and 2) is that 1) is a straight DH mapping, while 2) is the custom module.

Please let me know if what I said makes sense.  Thank you!

**Comment by dmcassel** [comment]
@chenfiona line 2 is: 

    const testUtils = require("../../marklogic/testUtils");

I'm not sure what this refers to. When I import MarkLogic Unit Test's assertion functions, I'm importing `test-helper.xqy`. I also noticed that your module is a `.js` file, not a `.sjs`. Is this running in MarkLogic or in a different tier? 

Line 4 -- I just googled and found the JavaScript Jest library. Doesn't look like your code (as shared) is using it, though (no references to `datahub`). 

For marklogic-unit-test, I'd expect to see something like:

```javascript
  const test = require('/test/test-helper.xqy');
  ...
  test.assertJsonEqual(expected, result)
```

**Comment by chenfiona** [comment]
The .js can be called .sjs. 
1) The tester uses Jest for JS script testing. The tester will mock out the modules outside of MarkLogic. So the tester expects the Unit test framework independent of MarkLogic.
2) You can import test-helper within MarkLogic. I am ok with that.

**Comment by rjrudin** [comment]
Assuming conversation has been resolved, as it's been a year
